### DISCUSSION TITLE
Livewire Model Properties Turn to Arrays

### DISCUSSION THREAD

**Discussion by maweitzman** [livewire/livewire#4223] [1. Help]
I'm sure I'm just naive, but I can't for the life of me figure out why my Eloquent models/collections are being turned to arrays when storing them as public properties. We started using Livewire with Alpine.js about a month ago, and I LOVE so much about it. Can someone please help me understand this?

**Comment by joshhanley** [comment]
@maweitzman please share your component class and blade view as well as any other relevant code.

Thanks!

**Reply by maweitzman** [reply]
@joshhanley Hmmmm... well we are using a lot of nested components, but from the very most parent level, it is an issue. So in our codebase we have requests -- they represent needs submitted by agency workers on behalf of children and/or families in need. So here is the view that iterates over all the request objects on a page and renders a Card (which is the parent most Livewire component) for each one:

![Screen Shot 2021-11-16 at 2 18 59 PM](https://user-images.githubusercontent.com/17768713/142059390-473b1a26-e5ea-45d2-85ac-c644e9be3e60.png)

Now, here is the class for the Livewire Card component:

![carbon](https://user-images.githubusercontent.com/17768713/142059790-85098d31-037d-40b0-9616-5bcfade9addc.png)

Finally, here is the Blade view for the Livewire Card component:

![carbon (1)](https://user-images.githubusercontent.com/17768713/142060054-2f18a408-d6e8-43a3-bf07-d3f603ffccab.png)

Lastly, here is a picture of what I see in the Livewire extension for Chrome developer tools -- it shows that the petition object that was passed into the Card component has now been turned into an empty array. Any help would be a HUGE blessing!

![Screen Shot 2021-11-16 at 2 16 21 PM](https://user-images.githubusercontent.com/17768713/142059198-d578778e-c46c-44ff-b19f-1f94373fcada.png)

**Comment by joshhanley** [comment]
@maweitzman thanks for sharing! 

So a couple of things:

Firstly the Livewire extension is in Javascript, so has no sense of what a PHP object is, so my question for you is, is the petition actually being converted to an array in PHP?

You can check this by doing a `dump($this->petition);` (or even better use Ray app) in your render method and then interact with the page, to make sure it is still a php object after a request has been made.

The other main thing is, for security reasons, Livewire will not make any eloquent model properties available to the front end unless you explicitly add those properties to your `$rules` property in the component class.

See here for more details https://laravel-livewire.com/docs/2.x/properties#binding-models

By adding rules, you then should see the properties appear in the Livewire extension.

Also for future reference, please just paste code directly instead of screenshots 🙂

When you paste code into github, you need to surround it in triple back ticks and then specify the file type like php or blade.

For example
\```php
\\ PHP code here
\```

\```blade
\\ HTML and blade code here
\```

Hope this helps!

**Reply by maweitzman** [reply]
@joshhanley thanks for tip -- `dump($this->petition)` showed that it is still an object -- which is good news.

So, here's the issue that actually led me to start the discussion yesterday:

On `$this->petition` there is an attribute `$this->petition->needs` that is an Eloquent Collection of our `RequestNeed` model. Our parent `Card` component has five, equal-level children components -- each of them represents a different tab in the card. The tab in question is our `Help` component. To quickly understand our UI, please refer to this screenshot:

![Screen Shot 2021-11-17 at 8 37 12 AM](https://user-images.githubusercontent.com/17768713/142220839-55cbe82c-207c-4ee1-b786-54a21f6a3356.png)

In short, when a user checks one of the need items in the list on the left, we are getting the corresponding object from `$this->petition` and then adding it into a new collection `committedItems` which is a class property. When a need item is unchecked, the corresponding object is removed from `$this->committedItems`. The crazy thing is (and I've confirmed this using the `dump()` trick you suggested) that when I add the first object to `$this->committedItems` it remains an object, but when I add a second (or anything higher) to the collection, it turns the recently added one into an array.

Where this becomes an issue is whenever a user clicks the 'Respond' button in the bottom right corner. That uses an Alpine.js trigger to open a confirmation modal that shows the list of `committedItems`. In the `@foreach` loop that iterates over `committedItems` and  then renders a child component for each (passing in the item object that corresponds with the current index of the loop) the item is already turned into an array before being passed to the child component.

Is this another Livewire caveat or am I missing something or what? Below is the class and view file, which in this case, should be all you need to see I believe -- let me know if you need more.

Help Component Class:
```php
<?php

namespace App\Http\Livewire\Entities\Requests\Tabs;

use Livewire\Component;

class Help extends Component
{
    public $data;
    public $petition;
    public $isFunding;
    public $response;
    public $committedItems;

    protected function getListeners()
    {
        return [
            'switchHelpModes' . $this->petition->id => 'switchHelpModes',
            'handleNeedItemCheckedChange',
            'handleNeedItemMeasurementChange',
            'refreshHelp-' . $this->petition->id => '$refresh'
        ];
    }

    public function mount(): void
    {
        $this->fill([
            'petition' => $this->data['petition'],
            'isFunding' => false,
            'response' => '',
            'committedItems' => collect([])
        ]);
    }

    public function switchHelpModes($isFunding)
    {
        $this->isFunding = $isFunding;
    }

    public function handleNeedItemCheckedChange($id)
    {
        if ($this->committedItems->contains('id', $id)) {
            $this->committedItems = $this->committedItems->keyBy('id');
            $this->committedItems->pull($id);
        } else {
            $changedItem = $this->petition->needs->keyBy('id')->get($id);
            
            // this is for what to display in the quantity/amount inputs in the confirm modal
            $changedItem->responseMeasurement = 0;

            $this->committedItems->push($changedItem);

            $this->committedItems = $this->committedItems->keyBy('id');
            $this->committedItems = $this->committedItems->sort();
        }
    }

    public function handleNeedItemMeasurementChange($id, $value)
    {
        dd($id, $value);
    }

    public function render()
    {
        return view('livewire.entities.requests.tabs.help');
    }
}
```

Help Component View:
```html
<div
    id="help-{{ $petition->id }}-content"
    class="flex space-x-6 h-80 content"
    x-data="{
        confirmResponse: false,
        response: @entangle('response'),
        toggleConfirmResponse() { this.confirmResponse = ! this.confirmResponse }
    }"
>
    <livewire:entities.requests.partials.needs-list.needs-list :petition="$petition" class="w-1/3 h-full" checkable/>

    <livewire:entities.requests.partials.help-forms :petition="$petition" :isFunding="$isFunding" class="w-2/3 h-full"/>

    <x-core.modal x-show="confirmResponse">
        <x-slot name="title">{{ $isFunding ? 'Fund' : 'Provide' }} Items Confirmation</x-slot>

        <div>
            <p class="mb-4">
                By clicking CONFIRM, you are stating that you are willing to {{ $isFunding ? 'fund' : 'provide' }} the items
                listed below. If this does not match what you were intending to fund, click GO BACK to revise your commitment.
            </p>

            @foreach ($committedItems as $item)
                <span class="mb-1">
                    <livewire:entities.requests.partials.needs-list.item :item="$item" checkable confirm :wire:key="$item['id']"/>
                </span>
            @endforeach

            @if ($isFunding)
                <span>Total amount being funded (including processing fee of $2.50): $52.50</span>
            @endif
        </div>

        <x-slot name="action">
            <button>Confirm</button>
        </x-slot>
    </x-core.modal>
</div>

@once
    @push('component-scripts')
        <script>
            function switchHelpModes(id, isFunding) {
                const helpTab = document.getElementById('help-' + id + '-content'); // get the entire yes i can help tab content

                const additionComponent = helpTab.querySelector('#needs-list-addition-section-' + id); // get the addition component in the needs list
                const fundFormFields = helpTab.querySelector('#fund-form-fields-' + id); // get all the funding-related form fields

                const allFormFields = helpTab.querySelectorAll('input, textarea'); // get all the inputs and textareas in the yes i can help tab content
                const modalLinks = helpTab.querySelectorAll('button.link'); // get all the links that open modals in the yes i can help tab content
                const respondButtons = helpTab.querySelector('#respond-buttons-' + id).children; // get all the buttons for responding

                const allColorChangingElements = [...allFormFields, ...modalLinks, ...respondButtons]; // get all elements that need to change color when funding

                if (! isFunding) { // if switching to provide
                    Livewire.emit('switchHelpModes' + id, isFunding); // emit event to switch help modes in component class

                    window.elementToggling([additionComponent, fundFormFields], 'hidden'); // removes all necessary elements from view

                    window.elementToggling(allColorChangingElements, 'fund', true); // changes colors green --> orange for all necessary elements

                    fundFormFields.reset(); // reset the fund form fields when switching back to Provide
                } else { // if switching to fund
                    Livewire.emit('switchHelpModes' + id, isFunding); // emit event to switch help modes in component class

                    window.elementToggling([additionComponent, fundFormFields], 'hidden', true); // adds all necessary elements to the view

                    window.elementToggling(allColorChangingElements, 'fund'); // changes colors orange --> green for all necessary elements
                }
            }
        </script>
    @endpush
@endonce
```

**Reply by joshhanley** [reply]
@maweitzman no worries! 

So what the issue is, is this part here in your mount method

```php
'committedItems' => collect([])
```

Livewire only knows how to rehydrate eloquent models if they are stand alone or in an eloquent collection specifically. But the `collect()` helper returns a standard collection.

So if you change that to 

```php
use Illuminate\Database\Eloquent\Collection as EloquentCollection;

'committedItems' => new EloquentCollection()
```

Then your models should be hydrated properly.

Hope this helps!
